File: go/boundary.go
package proofbundle

import (
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"strconv"
	"time"
)

var (
	ErrBoundaryEval = errors.New("boundary evaluation error")
)

// EvaluateBoundary evaluates a boundary expression against a context.
func EvaluateBoundary(boundaryJSON json.RawMessage, ctx map[string]interface{}, refTime time.Time) (bool, error) {
	if len(boundaryJSON) == 0 {
		return true, nil
	}
	var expr map[string]interface{}
	if err := json.Unmarshal(boundaryJSON, &expr); err != nil {
		return false, fmt.Errorf("%w: %v", ErrBoundaryEval, err)
	}
	return EvaluateAtom(expr, ctx, refTime)
}

// EvaluateAtom evaluates a single boundary atom.
func EvaluateAtom(atom map[string]interface{}, ctx map[string]interface{}, refTime time.Time) (bool, error) {
	if len(atom) != 1 {
		return false, fmt.Errorf("%w: atom must have exactly one key, got %v", ErrBoundaryEval, atom)
	}
	for op, args := range atom {
		switch op {
		case "equals":
			return evalEquals(args, ctx)
		case "in":
			return evalIn(args, ctx)
		case "range":
			return evalRange(args, ctx)
		case "present":
			return evalPresent(args, ctx)
		case "before":
			return evalBefore(args, ctx, refTime)
		case "after":
			return evalAfter(args, ctx, refTime)
		case "within":
			return evalWithin(args, ctx, refTime)
		case "expired":
			return evalExpired(args, ctx, refTime)
		case "not_expired":
			return evalNotExpired(args, ctx, refTime)
		case "age_lt":
			return evalAgeLT(args, ctx, refTime)
		case "age_gt":
			return evalAgeGT(args, ctx, refTime)
		case "within_last":
			return evalWithinLast(args, ctx, refTime)
		case "within_next":
			return evalWithinNext(args, ctx, refTime)
		case "all":
			return evalAll(args, ctx, refTime)
		case "any":
			return evalAny(args, ctx, refTime)
		case "not":
			return evalNot(args, ctx, refTime)
		default:
			return false, fmt.Errorf("%w: unknown boundary operator %q", ErrBoundaryEval, op)
		}
	}
	return false, fmt.Errorf("%w: empty atom", ErrBoundaryEval)
}

func toSlice(v interface{}) ([]interface{}, bool) {
	switch val := v.(type) {
	case []interface{}:
		return val, true
	default:
		return nil, false
	}
}

func toString(v interface{}) (string, bool) {
	switch val := v.(type) {
	case string:
		return val, true
	default:
		return "", false
	}
}

func toFloat(v interface{}) (float64, bool) {
	switch val := v.(type) {
	case float64:
		return val, true
	case int:
		return float64(val), true
	case int64:
		return float64(val), true
	case json.Number:
		n, err := val.Float64()
		return n, err == nil
	default:
		return 0, false
	}
}

func ctxValue(ctx map[string]interface{}, key string) (interface{}, bool) {
	v, ok := ctx[key]
	return v, ok
}

func evalEquals(args interface{}, ctx map[string]interface{}) (bool, error) {
	sl, ok := toSlice(args)
	if !ok || len(sl) != 2 {
		return false, fmt.Errorf("%w: equals requires [field, value] array", ErrBoundaryEval)
	}
	field, ok := toString(sl[0])
	if !ok {
		return false, fmt.Errorf("%w: equals field must be string", ErrBoundaryEval)
	}
	ctxVal, ok := ctxValue(ctx, field)
	if !ok {
		return false, nil
	}
	return valuesEqual(ctxVal, sl[1]), nil
}

func evalIn(args interface{}, ctx map[string]interface{}) (bool, error) {
	sl, ok := toSlice(args)
	if !ok || len(sl) != 2 {
		return false, fmt.Errorf("%w: in requires [field, values] array", ErrBoundaryEval)
	}
	field, ok := toString(sl[0])
	if !ok {
		return false, fmt.Errorf("%w: in field must be string", ErrBoundaryEval)
	}
	ctxVal, ok := ctxValue(ctx, field)
	if !ok {
		return false, nil
	}
	list, ok := toSlice(sl[1])
	if !ok {
		return false, fmt.Errorf("%w: in values must be array", ErrBoundaryEval)
	}
	for _, v := range list {
		if valuesEqual(ctxVal, v) {
			return true, nil
		}
	}
	return false, nil
}

func evalRange(args interface{}, ctx map[string]interface{}) (bool, error) {
	sl, ok := toSlice(args)
	if !ok || len(sl) != 3 {
		return false, fmt.Errorf("%w: range requires [field, min, max] array", ErrBoundaryEval)
	}
	field, ok := toString(sl[0])
	if !ok {
		return false, fmt.Errorf("%w: range field must be string", ErrBoundaryEval)
	}
	ctxVal, ok := ctxValue(ctx, field)
	if !ok {
		return false, nil
	}
	v, ok := toFloat(ctxVal)
	if !ok {
		return false, nil
	}
	min, ok := toFloat(sl[1])
	if !ok {
		return false, fmt.Errorf("%w: range min must be number", ErrBoundaryEval)
	}
	max, ok := toFloat(sl[2])
	if !ok {
		return false, fmt.Errorf("%w: range max must be number", ErrBoundaryEval)
	}
	return v >= min && v <= max, nil
}

func evalPresent(args interface{}, ctx map[string]interface{}) (bool, error) {
	field, ok := toString(args)
	if !ok {
		return false, fmt.Errorf("%w: present requires string field name", ErrBoundaryEval)
	}
	_, ok = ctxValue(ctx, field)
	return ok, nil
}

func parseTime(v interface{}) (time.Time, error) {
	s, ok := toString(v)
	if !ok {
		return time.Time{}, fmt.Errorf("expected string time, got %T", v)
	}
	for _, layout := range []string{
		time.RFC3339Nano,
		time.RFC3339,
		"2006-01-02T15:04:05Z07:00",
		"2006-01-02",
	} {
		t, err := time.Parse(layout, s)
		if err == nil {
			return t, nil
		}
	}
	return time.Time{}, fmt.Errorf("unparseable time: %q", s)
}

func ctxTime(ctx map[string]interface{}, field string) (time.Time, bool, error) {
	v, ok := ctx[field]
	if !ok {
		return time.Time{}, false, nil
	}
	t, err := parseTime(v)
	if err != nil {
		return time.Time{}, false, err
	}
	return t, true, nil
}

func evalBefore(args interface{}, ctx map[string]interface{}, refTime time.Time) (bool, error) {
	sl, ok := toSlice(args)
	if !ok || len(sl) != 2 {
		return false, fmt.Errorf("%w: before requires [field, time] array", ErrBoundaryEval)
	}
	field, ok := toString(sl[0])
	if !ok {
		return false, fmt.Errorf("%w: before field must be string", ErrBoundaryEval)
	}
	ctxT, found, err := ctxTime(ctx, field)
	if err != nil {
		return false, err
	}
	if !found {
		return false, nil
	}
	bound, err := parseTime(sl[1])
	if err != nil {
		return false, err
	}
	return ctxT.Before(bound), nil
}

func evalAfter(args interface{}, ctx map[string]interface{}, refTime time.Time) (bool, error) {
	sl, ok := toSlice(args)
	if !ok || len(sl) != 2 {
		return false, fmt.Errorf("%w: after requires [field, time] array", ErrBoundaryEval)
	}
	field, ok := toString(sl[0])
	if !ok {
		return false, fmt.Errorf("%w: after field must be string", ErrBoundaryEval)
	}
	ctxT, found, err := ctxTime(ctx, field)
	if err != nil {
		return false, err
	}
	if !found {
		return false, nil
	}
	bound, err := parseTime(sl[1])
	if err != nil {
		return false, err
	}
	return ctxT.After(bound), nil
}

func evalWithin(args interface{}, ctx map[string]interface{}, refTime time.Time) (bool, error) {
	sl, ok := toSlice(args)
	if !ok || len(sl) != 3 {
		return false, fmt.Errorf("%w: within requires [field, start, end] array", ErrBoundaryEval)
	}
	field, ok := toString(sl[0])
	if !ok {
		return false, fmt.Errorf("%w: within field must be string", ErrBoundaryEval)
	}
	ctxT, found, err := ctxTime(ctx, field)
	if err != nil {
		return false, err
	}
	if !found {
		return false, nil
	}
	start, err := parseTime(sl[1])
	if err != nil {
		return false, err
	}
	end, err := parseTime(sl[2])
	if err != nil {
		return false, err
	}
	return (ctxT.Equal(start) || ctxT.After(start)) && (ctxT.Equal(end) || ctxT.Before(end)), nil
}

func evalExpired(args interface{}, ctx map[string]interface{}, refTime time.Time) (bool, error) {
	field, ok := toString(args)
	if !ok {
		return false, fmt.Errorf("%w: expired requires string field name", ErrBoundaryEval)
	}
	ctxT, found, err := ctxTime(ctx, field)
	if err != nil {
		return false, err
	}
	if !found {
		return false, nil
	}
	return refTime.After(ctxT), nil
}

func evalNotExpired(args interface{}, ctx map[string]interface{}, refTime time.Time) (bool, error) {
	field, ok := toString(args)
	if !ok {
		return false, fmt.Errorf("%w: not_expired requires string field name", ErrBoundaryEval)
	}
	ctxT, found, err := ctxTime(ctx, field)
	if err != nil {
		return false, err
	}
	if !found {
		return false, nil
	}
	return !refTime.After(ctxT), nil
}

func evalAgeLT(args interface{}, ctx map[string]interface{}, refTime time.Time) (bool, error) {
	sl, ok := toSlice(args)
	if !ok || len(sl) != 2 {
		return false, fmt.Errorf("%w: age_lt requires [field, seconds] array", ErrBoundaryEval)
	}
	field, ok := toString(sl[0])
	if !ok {
		return false, fmt.Errorf("%w: age_lt field must be string", ErrBoundaryEval)
	}
	limit, ok := toFloat(sl[1])
	if !ok || math.IsNaN(limit) {
		return false, fmt.Errorf("%w: age_lt seconds must be number", ErrBoundaryEval)
	}
	ctxT, found, err := ctxTime(ctx, field)
	if err != nil {
		return false, err
	}
	if !found {
		return false, nil
	}
	age := refTime.Sub(ctxT).Seconds()
	return age < limit, nil
}

func evalAgeGT(args interface{}, ctx map[string]interface{}, refTime time.Time) (bool, error) {
	sl, ok := toSlice(args)
	if !ok || len(sl) != 2 {
		return false, fmt.Errorf("%w: age_gt requires [field, seconds] array", ErrBoundaryEval)
	}
	field, ok := toString(sl[0])
	if !ok {
		return false, fmt.Errorf("%w: age_gt field must be string", ErrBoundaryEval)
	}
	limit, ok := toFloat(sl[1])
	if !ok || math.IsNaN(limit) {
		return false, fmt.Errorf("%w: age_gt seconds must be number", ErrBoundaryEval)
	}
	ctxT, found, err := ctxTime(ctx, field)
	if err != nil {
		return false, err
	}
	if !found {
		return false, nil
	}
	age := refTime.Sub(ctxT).Seconds()
	return age > limit, nil
}

func evalWithinLast(args interface{}, ctx map[string]interface{}, refTime time.Time) (bool, error) {
	sl, ok := toSlice(args)
	if !ok || len(sl) != 2 {
		return false, fmt.Errorf("%w: within_last requires [field, seconds] array", ErrBoundaryEval)
	}
	field, ok := toString(sl[0])
	if !ok {
		return false, fmt.Errorf("%w: within_last field must be string", ErrBoundaryEval)
	}
	limit, ok := toFloat(sl[1])
	if !ok || math.IsNaN(limit) {
		return false, fmt.Errorf("%w: within_last seconds must be number", ErrBoundaryEval)
	}
	ctxT, found, err := ctxTime(ctx, field)
	if err != nil {
		return false, err
	}
	if !found {
		return false, nil
	}
	age := refTime.Sub(ctxT).Seconds()
	return age >= 0 && age <= limit, nil
}

func evalWithinNext(args interface{}, ctx map[string]interface{}, refTime time.Time) (bool, error) {
	sl, ok := toSlice(args)
	if !ok || len(sl) != 2 {
		return false, fmt.Errorf("%w: within_next requires [field, seconds] array", ErrBoundaryEval)
	}
	field, ok := toString(sl[0])
	if !ok {
		return false, fmt.Errorf("%w: within_next field must be string", ErrBoundaryEval)
	}
	limit, ok := toFloat(sl[1])
	if !ok || math.IsNaN(limit) {
		return false, fmt.Errorf("%w: within_next seconds must be number", ErrBoundaryEval)
	}
	ctxT, found, err := ctxTime(ctx, field)
	if err != nil {
		return false, err
	}
	if !found {
		return false, nil
	}
	diff := ctxT.Sub(refTime).Seconds()
	return diff >= 0 && diff <= limit, nil
}

func evalAll(args interface{}, ctx map[string]interface{}, refTime time.Time) (bool, error) {
	sl, ok := toSlice(args)
	if !ok {
		return false, fmt.Errorf("%w: all requires array of atoms", ErrBoundaryEval)
	}
	for _, item := range sl {
		atom, ok := item.(map[string]interface{})
		if !ok {
			return false, fmt.Errorf("%w: all item must be object", ErrBoundaryEval)
		}
		result, err := EvaluateAtom(atom, ctx, refTime)
		if err != nil {
			return false, err
		}
		if !result {
			return false, nil
		}
	}
	return true, nil
}

func evalAny(args interface{}, ctx map[string]interface{}, refTime time.Time) (bool, error) {
	sl, ok := toSlice(args)
	if !ok {
		return false, fmt.Errorf("%w: any requires array of atoms", ErrBoundaryEval)
	}
	for _, item := range sl {
		atom, ok := item.(map[string]interface{})
		if !ok {
			return false, fmt.Errorf("%w: any item must be object", ErrBoundaryEval)
		}
		result, err := EvaluateAtom(atom, ctx, refTime)
		if err != nil {
			return false, err
		}
		if result {
			return true, nil
		}
	}
	return false, nil
}

func evalNot(args interface{}, ctx map[string]interface{}, refTime time.Time) (bool, error) {
	atom, ok := args.(map[string]interface{})
	if !ok {
		return false, fmt.Errorf("%w: not requires object atom", ErrBoundaryEval)
	}
	result, err := EvaluateAtom(atom, ctx, refTime)
	if err != nil {
		return false, err
	}
	return !result, nil
}

func valuesEqual(a, b interface{}) bool {
	sa, aIsStr := toString(a)
	sb, bIsStr := toString(b)
	if aIsStr && bIsStr {
		return sa == sb
	}
	fa, aIsNum := toFloat(a)
	fb, bIsNum := toFloat(b)
	if aIsNum && bIsNum {
		return fa == fb
	}
	if aIsStr && bIsNum {
		// Try parse string as number
		if n, err := strconv.ParseFloat(sa, 64); err == nil {
			return n == fb
		}
	}
	if aIsNum && bIsStr {
		if n, err := strconv.ParseFloat(sb, 64); err == nil {
			return fa == n
		}
	}
	// JSON deep equality for arrays/objects
	ba, _ := json.Marshal(a)
	bb, _ := json.Marshal(b)
	return string(ba) == string(bb)
}

File: cli/src/boundary.rs
//! Boundary atom evaluator for ProofBundle.
//!
//! Atoms: equals, in, range, present, before, after, within, expired,
//!        not_expired, age_lt, age_gt, within_last, within_next,
//!        all, any, not

use serde_json::Value;

#[derive(Debug, Clone, PartialEq, thiserror::Error)]
pub enum BoundaryError {
    #[error("Unknown atom type: {0}")]
    UnknownAtom(String),
    #[error("Invalid atom arguments: {0}")]
    InvalidArguments(String),
    #[error("Type mismatch in atom evaluation: {0}")]
    TypeMismatch(String),
}

/// Evaluate a boundary atom against a value context.
///
/// The `context` is the JSON value against which the boundary is evaluated.
/// The `atom` is the boundary specification JSON.
pub fn evaluate(context: &Value, atom: &Value) -> Result<bool, BoundaryError> {
    match atom {
        Value::Object(map) => {
            let atom_type = map
                .get("type")
                .and_then(|v| v.as_str())
                .ok_or_else(|| BoundaryError::InvalidArguments("atom missing 'type'".to_string()))?;
            evaluate_atom(context, atom_type, map)
        }
        _ => Err(BoundaryError::InvalidArguments(
            "atom must be an object".to_string(),
        )),
    }
}

fn evaluate_atom(
    context: &Value,
    atom_type: &str,
    args: &serde_json::Map<String, Value>,
) -> Result<bool, BoundaryError> {
    match atom_type {
        // Field atoms: operate on a specific field from context
        "equals" => eval_equals(context, args),
        "in" => eval_in(context, args),
        "range" => eval_range(context, args),
        "present" => eval_present(context, args),
        "before" => eval_before(context, args),
        "after" => eval_after(context, args),
        "within" => eval_within(context, args),
        "expired" => eval_expired(context, args),
        "not_expired" => eval_not_expired(context, args),
        "age_lt" => eval_age_lt(context, args),
        "age_gt" => eval_age_gt(context, args),
        "within_last" => eval_within_last(context, args),
        "within_next" => eval_within_next(context, args),

        // Logical atoms
        "all" => eval_all(context, args),
        "any" => eval_any(context, args),
        "not" => eval_not(context, args),

        other => Err(BoundaryError::UnknownAtom(other.to_string())),
    }
}

fn get_field<'a>(context: &'a Value, args: &serde_json::Map<String, Value>) -> Option<&'a Value> {
    args.get("field")
        .and_then(|f| f.as_str())
        .and_then(|field_name| context.get(field_name))
}

fn eval_equals(context: &Value, args: &serde_json::Map<String, Value>) -> Result<bool, BoundaryError> {
    let value = get_field(context, args).unwrap_or(&Value::Null);
    let expected = args
        .get("value")
        .ok_or_else(|| BoundaryError::InvalidArguments("equals: missing 'value'".to_string()))?;
    Ok(value == expected)
}

fn eval_in(context: &Value, args: &serde_json::Map<String, Value>) -> Result<bool, BoundaryError> {
    let value = get_field(context, args).unwrap_or(&Value::Null);
    let list = args
        .get("values")
        .and_then(|v| v.as_array())
        .ok_or_else(|| BoundaryError::InvalidArguments("in: missing 'values' array".to_string()))?;
    Ok(list.contains(value))
}

fn eval_range(
    context: &Value,
    args: &serde_json::Map<String, Value>,
) -> Result<bool, BoundaryError> {
    let value = get_field(context, args).unwrap_or(&Value::Null);
    let min = args.get("min");
    let max = args.get("max");
    if min.is_none() && max.is_none() {
        return Err(BoundaryError::InvalidArguments(
            "range: need at least 'min' or 'max'".to_string(),
        ));
    }
    let meets_min = match min {
        Some(min_v) => compare_values(value, min_v).map(|c| c >= 0)?,
        None => true,
    };
    let meets_max = match max {
        Some(max_v) => compare_values(value, max_v).map(|c| c <= 0)?,
        None => true,
    };
    Ok(meets_min && meets_max)
}

fn eval_present(
    context: &Value,
    args: &serde_json::Map<String, Value>,
) -> Result<bool, BoundaryError> {
    match args.get("field").and_then(|f| f.as_str()) {
        Some(field) => {
            let present = context.get(field).is_some();
            Ok(present)
        }
        None => Err(BoundaryError::InvalidArguments(
            "present: missing 'field'".to_string(),
        )),
    }
}

fn eval_before(
    context: &Value,
    args: &serde_json::Map<String, Value>,
) -> Result<bool, BoundaryError> {
    let value = get_field(context, args).unwrap_or(&Value::Null);
    let bound = args
        .get("value")
        .ok_or_else(|| BoundaryError::InvalidArguments("before: missing 'value'".to_string()))?;
    compare_values(value, bound)
        .map(|c| c < 0)
        .map_err(|_| BoundaryError::TypeMismatch("before: cannot compare".to_string()))
}

fn eval_after(
    context: &Value,
    args: &serde_json::Map<String, Value>,
) -> Result<bool, BoundaryError> {
    let value = get_field(context, args).unwrap_or(&Value::Null);
    let bound = args
        .get("value")
        .ok_or_else(|| BoundaryError::InvalidArguments("after: missing 'value'".to_string()))?;
    compare_values(value, bound)
        .map(|c| c > 0)
        .map_err(|_| BoundaryError::TypeMismatch("after: cannot compare".to_string()))
}

fn eval_within(
    context: &Value,
    args: &serde_json::Map<String, Value>,
) -> Result<bool, BoundaryError> {
    let value = get_field(context, args).unwrap_or(&Value::Null);
    let min = args
        .get("min")
        .ok_or_else(|| BoundaryError::InvalidArguments("within: missing 'min'".to_string()))?;
    let max = args
        .get("max")
        .ok_or_else(|| BoundaryError::InvalidArguments("within: missing 'max'".to_string()))?;
    let ge_min = compare_values(value, min).map(|c| c >= 0).unwrap_or(false);
    let le_max = compare_values(value, max).map(|c| c <= 0).unwrap_or(false);
    Ok(ge_min && le_max)
}

fn eval_expired(
    context: &Value,
    args: &serde_json::Map<String, Value>,
) -> Result<bool, BoundaryError> {
    let now = chrono::Utc::now();
    let field_val = get_field(context, args).unwrap_or(&Value::Null);
    match field_val.as_str() {
        Some(ts) => {
            let dt = chrono::DateTime::parse_from_rfc3339(ts)
                .map_err(|e| BoundaryError::InvalidArguments(format!("expired: {e}")))?;
            Ok(now > dt.with_timezone(&chrono::Utc))
        }
        None => Ok(false),
    }
}

fn eval_not_expired(
    context: &Value,
    args: &serde_json::Map<String, Value>,
) -> Result<bool, BoundaryError> {
    eval_expired(context, args).map(|r| !r)
}

fn eval_age_lt(
    context: &Value,
    args: &serde_json::Map<String, Value>,
) -> Result<bool, BoundaryError> {
    let now = chrono::Utc::now();
    let field_val = get_field(context, args).unwrap_or(&Value::Null);
    let max_seconds = args
        .get("seconds")
        .and_then(|v| v.as_i64())
        .ok_or_else(|| BoundaryError::InvalidArguments("age_lt: missing 'seconds'".to_string()))?;
    match field_val.as_str() {
        Some(ts) => {
            let dt = chrono::DateTime::parse_from_rfc3339(ts)
                .map_err(|e| BoundaryError::InvalidArguments(format!("age_lt: {e}")))?;
            let age = now.signed_duration_since(dt.with_timezone(&chrono::Utc));
            Ok(age.num_seconds() < max_seconds)
        }
        None => Ok(false),
    }
}

fn eval_age_gt(
    context: &Value,
    args: &serde_json::Map<String, Value>,
) -> Result<bool, BoundaryError> {
    let now = chrono::Utc::now();
    let field_val = get_field(context, args).unwrap_or(&Value::Null);
    let min_seconds = args
        .get("seconds")
        .and_then(|v| v.as_i64())
        .ok_or_else(|| BoundaryError::InvalidArguments("age_gt: missing 'seconds'".to_string()))?;
    match field_val.as_str() {
        Some(ts) => {
            let dt = chrono::DateTime::parse_from_rfc3339(ts)
                .map_err(|e| BoundaryError::InvalidArguments(format!("age_gt: {e}")))?;
            let age = now.signed_duration_since(dt.with_timezone(&chrono::Utc));
            Ok(age.num_seconds() > min_seconds)
        }
        None => Ok(false),
    }
}

fn eval_within_last(
    context: &Value,
    args: &serde_json::Map<String, Value>,
) -> Result<bool, BoundaryError> {
    let now = chrono::Utc::now();
    let field_val = get_field(context, args).unwrap_or(&Value::Null);
    let seconds = args
        .get("seconds")
        .and_then(|v| v.as_i64())
        .ok_or_else(|| {
            BoundaryError::InvalidArguments("within_last: missing 'seconds'".to_string())
        })?;
    match field_val.as_str() {
        Some(ts) => {
            let dt = chrono::DateTime::parse_from_rfc3339(ts)
                .map_err(|e| BoundaryError::InvalidArguments(format!("within_last: {e}")))?;
            let age = now.signed_duration_since(dt.with_timezone(&chrono::Utc));
            Ok(age.num_seconds() >= 0 && age.num_seconds() <= seconds)
        }
        None => Ok(false),
    }
}

fn eval_within_next(
    context: &Value,
    args: &serde_json::Map<String, Value>,
) -> Result<bool, BoundaryError> {
    let now = chrono::Utc::now();
    let field_val = get_field(context, args).unwrap_or(&Value::Null);
    let seconds = args
        .get("seconds")
        .and_then(|v| v.as_i64())
        .ok_or_else(|| {
            BoundaryError::InvalidArguments("within_next: missing 'seconds'".to_string())
        })?;
    match field_val.as_str() {
        Some(ts) => {
            let dt = chrono::DateTime::parse_from_rfc3339(ts)
                .map_err(|e| BoundaryError::InvalidArguments(format!("within_next: {e}")))?;
            let diff = dt.with_timezone(&chrono::Utc).signed_duration_since(now);
            Ok(diff.num_seconds() >= 0 && diff.num_seconds() <= seconds)
        }
        None => Ok(false),
    }
}

fn eval_all(context: &Value, args: &serde_json::Map<String, Value>) -> Result<bool, BoundaryError> {
    let conditions = args
        .get("conditions")
        .and_then(|v| v.as_array())
        .ok_or_else(|| BoundaryError::InvalidArguments("all: missing 'conditions'".to_string()))?;
    for cond in conditions {
        if !evaluate(context, cond)? {
            return Ok(false);
        }
    }
    Ok(true)
}

fn eval_any(context: &Value, args: &serde_json::Map<String, Value>) -> Result<bool, BoundaryError> {
    let conditions = args
        .get("conditions")
        .and_then(|v| v.as_array())
        .ok_or_else(|| BoundaryError::InvalidArguments("any: missing 'conditions'".to_string()))?;
    if conditions.is_empty() {
        return Ok(false);
    }
    for cond in conditions {
        if evaluate(context, cond)? {
            return Ok(true);
        }
    }
    Ok(false)
}

fn eval_not(context: &Value, args: &serde_json::Map<String, Value>) -> Result<bool, BoundaryError> {
    let condition = args
        .get("condition")
        .ok_or_else(|| BoundaryError::InvalidArguments("not: missing 'condition'".to_string()))?;
    evaluate(context, condition).map(|r| !r)
}

/// Compare two JSON values. Returns:
/// - Ok(-1) if a < b
/// - Ok(0) if a == b
/// - Ok(1) if a > b
fn compare_values(a: &Value, b: &Value) -> Result<i32, BoundaryError> {
    match (a, b) {
        (Value::Null, Value::Null) => Ok(0),
        (Value::Bool(a), Value::Bool(b)) => {
            if a == b {
                Ok(0)
            } else if *a {
                Ok(1)
            } else {
                Ok(-1)
            }
        }
        (Value::Number(a), Value::Number(b)) => {
            match (a.as_i64(), b.as_i64()) {
                (Some(ai), Some(bi)) => {
                    if ai < bi {
                        Ok(-1)
                    } else if ai > bi {
                        Ok(1)
                    } else {
                        Ok(0)
                    }
                }
                _ => Err(BoundaryError::TypeMismatch(
                    "compare: non-integer numbers".to_string(),
                )),
            }
        }
        (Value::String(a), Value::String(b)) => Ok(a.cmp(b) as i32),
        _ => Err(BoundaryError::TypeMismatch(
            "compare: incompatible types".to_string(),
        )),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_equals() {
        let ctx = serde_json::json!({"role": "admin"});
        let atom = serde_json::json!({"type": "equals", "field": "role", "value": "admin"});
        assert!(evaluate(&ctx, &atom).unwrap());

        let atom2 = serde_json::json!({"type": "equals", "field": "role", "value": "user"});
        assert!(!evaluate(&ctx, &atom2).unwrap());
    }

    #[test]
    fn test_in() {
        let ctx = serde_json::json!({"role": "admin"});
        let atom = serde_json::json!({"type": "in", "field": "role", "values": ["admin", "user"]});
        assert!(evaluate(&ctx, &atom).unwrap());
    }

    #[test]
    fn test_range() {
        let ctx = serde_json::json!({"age": 25});
        let atom = serde_json::json!({"type": "range", "field": "age", "min": 18, "max": 65});
        assert!(evaluate(&ctx, &atom).unwrap());
    }

    #[test]
    fn test_present() {
        let ctx = serde_json::json!({"name": "test"});
        let atom = serde_json::json!({"type": "present", "field": "name"});
        assert!(evaluate(&ctx, &atom).unwrap());

        let atom2 = serde_json::json!({"type": "present", "field": "missing"});
        assert!(!evaluate(&ctx, &atom2).unwrap());
    }

    #[test]
    fn test_all() {
        let ctx = serde_json::json!({"role": "admin", "active": true});
        let atom = serde_json::json!({
            "type": "all",
            "conditions": [
                {"type": "equals", "field": "role", "value": "admin"},
                {"type": "equals", "field": "active", "value": true}
            ]
        });
        assert!(evaluate(&ctx, &atom).unwrap());
    }

    #[test]
    fn test_any() {
        let ctx = serde_json::json!({"role": "user"});
        let atom = serde_json::json!({
            "type": "any",
            "conditions": [
                {"type": "equals", "field": "role", "value": "admin"},
                {"type": "equals", "field": "role", "value": "user"}
            ]
        });
        assert!(evaluate(&ctx, &atom).unwrap());
    }

    #[test]
    fn test_not() {
        let ctx = serde_json::json!({"role": "user"});
        let atom = serde_json::json!({
            "type": "not",
            "condition": {"type": "equals", "field": "role", "value": "admin"}
        });
        assert!(evaluate(&ctx, &atom).unwrap());
    }
}
